From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 18 Apr 2026 10:50:29 +0800
Subject: [PATCH] Add back BlockPos transform to Position support


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 08367eaa9f93d6213502d76be2acfe95c723c556..d2cfdae7f32d29cddf4dda242749a7d7142ec1c1 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -168,6 +168,12 @@ public final class MCUtil {
         return CraftLocation.toBukkit(pos, world);
     }
 
+    // Luminol start - add BlockPos support
+    public static Location toLocation(Level world, BlockPos pos) {
+        return CraftLocation.toBukkit(pos, world);
+    }
+    // Luminol end - add BlockPos support
+
     /**
      * Converts a NMS World/Vector to Bukkit Location
      *
