From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 20 Apr 2026 01:10:30 +0800
Subject: [PATCH] Fix entity portal-teleport speed


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index a51a9e7a0fa5727690046d814eb0b26d5742cd74..fbcbaf1840a1ee4f354da0245e9cf5708ef5f84e 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1503,7 +1503,27 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         foliaProfiler.startTimer(timerId);
         try {
         // Folia end - profiler
+        // Luminol start - Entity portal-teleport speed fix
         if (isActive) { // Paper - EAR 2
+            if (!(entity instanceof Player) && entity.teleportTickType == 2) { // Luminol - after portal compensate tick
+                entity.tick();
+                entity.tick();
+                entity.teleportTickType = 0;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else if (!(entity instanceof Player) && entity.teleportTickType == 1) { // Luminol - portal teleport only
+                entity.teleportTickType++;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else {
         entity.tick();
         // Folia start - region threading
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
@@ -1514,6 +1534,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             // portalled
             return;
         }
+        }
+        // Luminol end - Entity portal-teleport speed fix
         // Folia end - region threading
         } else {entity.inactiveTick();} // Paper - EAR 2
         profiler.pop();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 3417f86c9d5f16e91404a1252f72d5344b37ea34..ff7e282eda93e07184c7fc59c92ff735abd16bd3 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -387,6 +387,7 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public int teleportTickType = 0;// Luminol - Entity portal-teleport speed fix
 
     public void inactiveTick() {
     }
@@ -3530,6 +3531,7 @@ public abstract class Entity
         } else {
             if (this.portalProcess == null || !this.portalProcess.isSamePortal(portal)) {
                 this.portalProcess = new PortalProcessor(portal, pos.immutable());
+                this.teleportTickType = 1; // Luminol - Entity portal-teleport speed fix
             } else if (!this.portalProcess.isInsidePortalThisTick()) {
                 this.portalProcess.updateEntryPosition(pos.immutable());
                 this.portalProcess.setAsInsidePortalThisTick(true);
