From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Tue, 21 Apr 2026 01:28:05 +0800
Subject: [PATCH] Temporarily fix teleport yam and pitch


diff --git a/net/minecraft/server/commands/TeleportCommand.java b/net/minecraft/server/commands/TeleportCommand.java
index bfc012df8bf64cc9d2be36b2a4e97973468aed8f..f3b697eda0f47a2d1c64d7978ccdfb77fa226929 100644
--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -247,8 +247,10 @@ public class TeleportCommand {
             // Folia start - region threading
             if (true) {
                 Vec3 posFinal = new Vec3(x, y, z);
-                Float yawFinal = Float.valueOf(newYRot);
-                Float pitchFinal = Float.valueOf(newXRot);
+                //Float yawFinal = Float.valueOf(newYRot); // Luminol - fix teleport yaw issue
+                //Float pitchFinal = Float.valueOf(newXRot); // Luminol - fix teleport yaw issue
+                Float yawFinal = Float.valueOf(newYRot + victim.getYRot()); // Luminol - fix teleport yaw issue
+                Float pitchFinal = Float.valueOf(newXRot + victim.getXRot()); // Luminol - fix teleport pitch issue
                 victim.getBukkitEntity().taskScheduler.schedule((Entity nmsEntity) -> {
                     nmsEntity.stopRiding();
                     nmsEntity.teleportAsync(
