From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 18 Apr 2026 09:15:04 +0800
Subject: [PATCH] Add config gui support


diff --git a/net/minecraft/server/dialog/action/ParsedTemplate.java b/net/minecraft/server/dialog/action/ParsedTemplate.java
index 3d60e3e50297f4f657e01fbc5b1d144596b29b33..2774380f65a70aca7dcc781aea2a0a38f6dc46d8 100644
--- a/net/minecraft/server/dialog/action/ParsedTemplate.java
+++ b/net/minecraft/server/dialog/action/ParsedTemplate.java
@@ -13,7 +13,7 @@ public class ParsedTemplate {
     private final String raw;
     private final StringTemplate parsed;
 
-    private ParsedTemplate(final String raw, final StringTemplate parsed) {
+    public ParsedTemplate(final String raw, final StringTemplate parsed) { // Luminol - make public
         this.raw = raw;
         this.parsed = parsed;
     }
