From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Fri, 5 Jun 2026 15:09:29 +0800
Subject: [PATCH] Add config to enable tick command

only freeze/unfreeze/step/query can run when enabled

diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index 3416445888860746c15a978f401c87245609b64a..11cacb26ce260373476dad764014ba1852477831 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -338,6 +338,11 @@ public final class RegionizedServer {
         this.randomWalk();
          */
         ++this.tickCount;
+        // Luminol start - Add a config to enable tick command
+        if (me.earthme.luminol.config.modules.experiment.CommandConfig.tick) {
+            MinecraftServer.getServer().tickRateManager().tick();
+        }
+        // Luminol end - Add a config to enable tick command
         // expire invalid click command callbacks
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue((int)this.tickCount); // Paper // Folia - region threading - moved to global tick
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.DIALOG_CLICK_MANAGER.handleQueue((int)this.tickCount); // Paper // Folia - region threading - moved to global tick
@@ -511,7 +516,7 @@ public final class RegionizedServer {
     }
 
     private void tickTime(final ServerLevel world, final long tickCount) {
-        if (world.tickTime) {
+        if ((!me.earthme.luminol.config.modules.experiment.CommandConfig.tick || world.tickRateManager().runsNormally()) && world.tickTime) { // Luminol - Add a config to enable tick command
             world.serverLevelData.setGameTime(world.serverLevelData.getGameTime() + tickCount);
         }
     }
diff --git a/io/papermc/paper/threadedregions/TickRegionScheduler.java b/io/papermc/paper/threadedregions/TickRegionScheduler.java
index 047a20d0c0e7dd9f9ff4ee61a348c2010b89ba51..86c51506c518ee445d00f7eef36ee2c9adbd9b4c 100644
--- a/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -559,6 +559,11 @@ public final class TickRegionScheduler {
             try {
                 // next start isn't updated until the end of this tick
                 this.tickRegion(tickCount, tickStart, scheduledEnd);
+                // Luminol start - Add a config to enable tick command
+                if (me.earthme.luminol.config.modules.experiment.CommandConfig.tick) {
+                    MinecraftServer.getServer().tickRateManager().endTickWork();
+                }
+                // Luminol end - Add a config to enable tick command
             } catch (final Throwable thr) {
                 this.scheduler.regionFailed(this, false, thr);
                 // regionFailed will schedule a shutdown, so we should avoid letting this region tick further
diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index f90e31c757d4ade81d7a07ad8cac87eaaf52f057..4a81f07c3502949166768a59173afdfc596a36b6 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -259,7 +259,11 @@ public class Commands {
         TeleportCommand.register(this.dispatcher);
         TellRawCommand.register(this.dispatcher, context);
         //TestCommand.register(this.dispatcher, context); // Folia - region threading
-        //TickCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        // Luminol start - Add a config to enable tick command
+        if (me.earthme.luminol.config.modules.experiment.CommandConfig.tick) {
+            TickCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        }
+        // Luminol end - Add a config to enable tick command
         TimeCommand.register(this.dispatcher, context);
         TitleCommand.register(this.dispatcher, context);
         //TriggerCommand.register(this.dispatcher); // Folia - region threading - TODO later
diff --git a/net/minecraft/server/commands/TickCommand.java b/net/minecraft/server/commands/TickCommand.java
index 2dd0a8f2399a83b90356fcf0777bd1aec8eef5a9..cae943bae4701f74b46a49d68da83ae797841590 100644
--- a/net/minecraft/server/commands/TickCommand.java
+++ b/net/minecraft/server/commands/TickCommand.java
@@ -22,14 +22,14 @@ public class TickCommand {
             Commands.literal("tick")
                 .requires(Commands.hasPermission(Commands.LEVEL_ADMINS))
                 .then(Commands.literal("query").executes(c -> tickQuery(c.getSource())))
-                .then(
+/*                .then(
                     Commands.literal("rate")
                         .then(
                             Commands.argument("rate", FloatArgumentType.floatArg(1.0F, 10000.0F))
                                 .suggests((c, b) -> SharedSuggestionProvider.suggest(new String[]{DEFAULT_TICKRATE}, b))
                                 .executes(c -> setTickingRate(c.getSource(), FloatArgumentType.getFloat(c, "rate")))
                         )
-                )
+                )*/
                 .then(
                     Commands.literal("step")
                         .executes(c -> step(c.getSource(), 1))
@@ -40,7 +40,7 @@ public class TickCommand {
                                 .executes(c -> step(c.getSource(), IntegerArgumentType.getInteger(c, "time")))
                         )
                 )
-                .then(
+/*                .then(
                     Commands.literal("sprint")
                         .then(Commands.literal("stop").executes(c -> stopSprinting(c.getSource())))
                         .then(
@@ -48,7 +48,7 @@ public class TickCommand {
                                 .suggests((c, b) -> SharedSuggestionProvider.suggest(new String[]{"60s", "1d", "3d"}, b))
                                 .executes(c -> sprint(c.getSource(), IntegerArgumentType.getInteger(c, "time")))
                         )
-                )
+                )*/
                 .then(Commands.literal("unfreeze").executes(c -> setFreeze(c.getSource(), false)))
                 .then(Commands.literal("freeze").executes(c -> setFreeze(c.getSource(), true)))
         );
