From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 18 Apr 2026 10:50:12 +0800
Subject: [PATCH] Add missing teleportation event APIs for folia


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 451974ccf2dea148d5465d7fbb0f702e5260fd07..db42de1e67dddc9dfdcfffed3e2c43ae29c4ffe5 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1793,6 +1793,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         ca.spottedleaf.concurrentutil.completable.CallbackCompletable<org.bukkit.Location> spawnPosComplete =
             new ca.spottedleaf.concurrentutil.completable.CallbackCompletable<>();
         boolean[] usedRespawnAnchor = new boolean[1];
+        boolean[] postRespawnBedSpawn = new boolean[1];
+        boolean[] postRespawnAnchorSpawn = new boolean[1];
+        boolean[] postRespawnMissingBlock = new boolean[1];
 
         // set up post spawn location logic
         spawnPosComplete.addWaiter((spawnLoc, throwable) -> {
@@ -1829,6 +1832,16 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     if (finalRespawnCompleteCallback != null) { // Luminol - Level hot unload apis
                         finalRespawnCompleteCallback.accept(ServerPlayer.this); // Luminol - Level hot unload apis
                     }
+                    // Luminol - Add missing teleportation apis
+                    new me.earthme.luminol.api.entity.player.PostPlayerRespawnEvent(
+                            ServerPlayer.this.getBukkitEntity(),
+                            spawnLoc,
+                            postRespawnBedSpawn[0],
+                            postRespawnAnchorSpawn[0],
+                            postRespawnMissingBlock[0],
+                            reason
+                    ).callEvent();
+                    // Luminol end
                 }
             );
         });
@@ -1848,6 +1861,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                         respawnWorld, respawnConfig, !alive
                     ).orElse(null);
                     if (spawnPos == null) {
+                        postRespawnMissingBlock[0] = true;
                         // no spawn
                         ServerPlayer.this.connection.send(
                             new ClientboundGameEventPacket(ClientboundGameEventPacket.NO_RESPAWN_BLOCK_AVAILABLE, 0.0F)
@@ -1863,6 +1877,8 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     boolean isRespawnAnchor = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.world.level.block.Blocks.RESPAWN_ANCHOR);
                     boolean isBed = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.tags.BlockTags.BEDS);
                     usedRespawnAnchor[0] = !alive && isRespawnAnchor;
+                    postRespawnBedSpawn[0] = spawnPos.isBedSpawn();
+                    postRespawnAnchorSpawn[0] = spawnPos.isAnchorSpawn();
 
                     // finished now, pass the location on
                     spawnPosComplete.complete(
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 7c4f9a600183462f4e9bd31f9a4086bb0d653deb..b8ff929dd6825d87bd48e4b9ba95359e89e2f799 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4613,7 +4613,8 @@ public abstract class Entity
                                 new net.minecraft.world.level.portal.TeleportTransition(
                                     destination, finalPos, this.getDeltaMovement(), Direction.WEST.toYRot(), 0.0f, // Kaiiju - Vanilla end teleportation
                                     /*Relative.union(Relative.DELTA, Set.of(Relative.X_ROT))*/Set.of(), // Kaiiju - Vanilla end teleportation
-                                    TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET),
+                                    TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET)
+                                        .then(entity -> net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, targetPos.below(), true, null)),
                                     org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL
                                 )
                             );
@@ -4813,11 +4814,37 @@ public abstract class Entity
         if (!this.canPortalAsync(destination, takePassengers)) {
             return false;
         }
-
-        // Kaiiju start - sync end platform spawning & entity teleportation
-        final java.util.function.Consumer<Entity> tpComplete = type == PortalType.END && destination.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END ?
-                e -> {net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, ServerLevel.END_SPAWN_POINT.below(), true, null); if (teleportComplete != null) {teleportComplete.accept(e);}} : teleportComplete;
-        // Kaiiju end
+        // Luminol start - Add missing teleportation events
+        if (!new me.earthme.luminol.api.entity.PreEntityPortalEvent(
+                this.getBukkitEntity(),
+                io.papermc.paper.util.MCUtil.toLocation(this.level, portalPos),
+                destination.getWorld(),
+                type == PortalType.END ? org.bukkit.PortalType.ENDER : org.bukkit.PortalType.NETHER,
+                type == PortalType.END
+                        ? org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL
+                        : org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.NETHER_PORTAL
+        ).callEvent()) {
+            return false;
+        }
+        // Luminol end
+        final org.bukkit.Location luminol$portalFrom = io.papermc.paper.util.MCUtil.toLocation(this.level, portalPos); // Luminol - Add missing teleportation events
+        final org.bukkit.PortalType luminol$portalType = type == PortalType.END ? org.bukkit.PortalType.ENDER : org.bukkit.PortalType.NETHER; // Luminol - Add missing teleportation events
+        final org.bukkit.event.player.PlayerTeleportEvent.TeleportCause luminol$teleportCause = type == PortalType.END
+                ? org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL
+                : org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.NETHER_PORTAL; // Luminol - Add missing teleportation events
+
+        final java.util.function.Consumer<Entity> luminol$portalComplete = teleported -> { // Luminol - Add missing teleportation events
+            new me.earthme.luminol.api.entity.PostEntityPortalEvent(
+                teleported.getBukkitEntity(),
+                luminol$portalFrom,
+                teleported.getBukkitEntity().getLocation(),
+                luminol$portalType,
+                luminol$teleportCause
+            ).callEvent();
+            if (teleportComplete != null) {
+                teleportComplete.accept(teleported);
+            }
+        }; // Luminol - Add missing teleportation events
         Vec3 initialPosition = this.position();
         ChunkPos initialPositionChunk = new ChunkPos(
             ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(initialPosition),
@@ -4896,8 +4923,8 @@ public abstract class Entity
                     /*if (teleportComplete != null) {
                         teleportComplete.accept(teleported);
                     }*/
-                    if (tpComplete != null){
-                        tpComplete.accept(teleported);
+                    if (luminol$portalComplete != null){ // Luminol - Add missing teleportation events
+                        luminol$portalComplete.accept(teleported); // Luminol - Add missing teleportation events
                     }
                     // Kaiiju end
                 }
