From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LittleOvO <littleovo@example.com>
Date: Tue, 14 Jul 2026 21:07:12 +0800
Subject: [PATCH] Arbor: Replace packet processing queue with mt-queue


diff --git a/net/minecraft/network/PacketProcessor.java b/net/minecraft/network/PacketProcessor.java
index 24f5b890b0126fc817863f99b676da1a548f7608..13ce2ae43b7b583bdad35d7a60eefbb65ba140f9 100644
--- a/net/minecraft/network/PacketProcessor.java
+++ b/net/minecraft/network/PacketProcessor.java
@@ -11,7 +11,7 @@ import org.slf4j.Logger;
 
 public class PacketProcessor implements AutoCloseable {
     private static final Logger LOGGER = LogUtils.getLogger();
-    private final Queue<PacketProcessor.ListenerAndPacket<?>> packetsToBeHandled = Queues.newConcurrentLinkedQueue();
+    private final Queue<PacketProcessor.ListenerAndPacket<?>> packetsToBeHandled = new ca.spottedleaf.concurrentutil.collection.MultiThreadedQueue<>();
     private final Thread runningThread;
     private boolean closed;
 
